File: examples/_go/memory.go
package main

import (
	"github.com/grafana/cog/generated/go/common"
	"github.com/grafana/cog/generated/go/dashboard"
	"github.com/grafana/cog/generated/go/gauge"
	"github.com/grafana/cog/generated/go/timeseries"
)

func memoryUsageTimeseries() *timeseries.PanelBuilder {
	memUsedQuery := `(
  node_memory_MemTotal_bytes{job="integrations/raspberrypi-node", instance="$instance"}
-
  node_memory_MemFree_bytes{job="integrations/raspberrypi-node", instance="$instance"}
-
  node_memory_Buffers_bytes{job="integrations/raspberrypi-node", instance="$instance"}
-
  node_memory_Cached_bytes{job="integrations/raspberrypi-node", instance="$instance"}
)`

	return defaultTimeseries().
		Title("Memory Usage").
		Span(18).
		Stacking(common.NewStackingConfigBuilder().Mode(common.StackingModeNormal)). // TODO: painful, not intuitive
		Thresholds(
			dashboard.NewThresholdsConfigBuilder().
				Mode(dashboard.ThresholdsModeAbsolute).
				Steps([]dashboard.Threshold{
					{Value: nil, Color: "green"},
					{Value: toPtr(80.0), Color: "red"},
				}),
		).
		Min(0).
		Unit("bytes").
		Decimals(2).
		WithTarget(
			basicPrometheusQuery(memUsedQuery, "Used"),
		).
		WithTarget(
			basicPrometheusQuery(`node_memory_Buffers_bytes{job="integrations/raspberrypi-node", instance="$instance"}`, "Buffers"),
		).
		WithTarget(
			basicPrometheusQuery(`node_memory_Cached_bytes{job="integrations/raspberrypi-node", instance="$instance"}`, "Cached"),
		).
		WithTarget(
			basicPrometheusQuery(`node_memory_MemFree_bytes{job="integrations/raspberrypi-node", instance="$instance"}`, "Free"),
		)
}

func memoryUsageGauge() *gauge.PanelBuilder {
	query := `100 - (
  avg(node_memory_MemAvailable_bytes{job="integrations/raspberrypi-node", instance="$instance"}) /
  avg(node_memory_MemTotal_bytes{job="integrations/raspberrypi-node", instance="$instance"})
* 100)`

	return defaultGauge().
		Title("Memory Usage").
		Span(6).
		Min(30).
		Max(100).
		Unit("percent").
		Thresholds(
			dashboard.NewThresholdsConfigBuilder().
				Mode(dashboard.ThresholdsModeAbsolute).
				Steps([]dashboard.Threshold{
					{Value: nil, Color: "rgba(50, 172, 45, 0.97)"},
					{Value: toPtr(80.0), Color: "rgba(237, 129, 40, 0.89)"},
					{Value: toPtr(90.0), Color: "rgba(245, 54, 54, 0.9)"},
				}),
		).
		WithTarget(
			basicPrometheusQuery(query, ""),
		)
}

File: examples/php/src/Monitoring/Memory.php
<?php

namespace App\Monitoring;

use Grafana\Foundation\Common\StackingConfigBuilder;
use Grafana\Foundation\Common\StackingMode;
use Grafana\Foundation\Dashboard\Threshold;
use Grafana\Foundation\Dashboard\ThresholdsConfigBuilder;
use Grafana\Foundation\Dashboard\ThresholdsMode;
use Grafana\Foundation\Gauge;
use Grafana\Foundation\Timeseries;

class Memory
{
    public static function usageTimeseries(): Timeseries\PanelBuilder
    {
        $query = <<<'PROMQL'
(
  node_memory_MemTotal_bytes{job="integrations/raspberrypi-node", instance="$instance"}
  -
  node_memory_MemFree_bytes{job="integrations/raspberrypi-node", instance="$instance"}
  -
  node_memory_Buffers_bytes{job="integrations/raspberrypi-node", instance="$instance"}
  -
  node_memory_Cached_bytes{job="integrations/raspberrypi-node", instance="$instance"}
)
PROMQL;
        
        return Common::defaultTimeseries()
            ->title('Memory Usage')
            ->span(18)
            ->stacking(
                (new StackingConfigBuilder())
                    ->mode(StackingMode::normal())
            )
            ->thresholds(
                (new ThresholdsConfigBuilder())
                    ->mode(ThresholdsMode::absolute())
                    ->steps([
                        new Threshold(color: 'green'),
                        new Threshold(color: 'red', value: 80.0),
                    ])
            )
            ->min(0)
            ->decimals(2)
            ->unit('bytes')
            ->withTarget(
                Common::basicPrometheusQuery($query, 'Used'),
            )
            ->withTarget(
                Common::basicPrometheusQuery('node_memory_Buffers_bytes{job="integrations/raspberrypi-node", instance="$instance"}', 'Buffers'),
            )
            ->withTarget(
                Common::basicPrometheusQuery('node_memory_Cached_bytes{job="integrations/raspberrypi-node", instance="$instance"}', 'Cached'),
            )
            ->withTarget(
                Common::basicPrometheusQuery('node_memory_MemFree_bytes{job="integrations/raspberrypi-node", instance="$instance"}', 'Free'),
            );
    }

    public static function usageGauge(): Gauge\PanelBuilder
    {
        $query = <<<'PROMQL'
100 - (
    avg(node_memory_MemAvailable_bytes{job="integrations/raspberrypi-node", instance="$instance"}) /
    avg(node_memory_MemTotal_bytes{job="integrations/raspberrypi-node", instance="$instance"})
    * 100
)
PROMQL;

        return Common::defaultGauge()
            ->title('Memory Usage')
            ->span(6)
            ->min(30)
            ->max(100)
            ->unit('percent')
            ->thresholds(
                (new ThresholdsConfigBuilder())
                    ->mode(ThresholdsMode::absolute())
                    ->steps([
                        new Threshold(color: 'rgba(50, 172, 45, 0.97)'),
                        new Threshold(color: 'rgba(237, 129, 40, 0.89)', value: 80.0),
                        new Threshold(color: 'rgba(245, 54, 54, 0.9)', value: 90.0),
                    ])
            )
            ->withTarget(
                Common::basicPrometheusQuery($query, ''),
            );
    }
}